ThunderCloud/StormLanguageController.swift

```diff
@@ -0,0 +1,446 @@
+//
+//  StormLanguageController.swift
+//  ThunderCloud
+//
+//  Created by Matthew Cheetham on 02/06/2017.
+//  Copyright © 2017 threesidedcube. All rights reserved.
+//
+
+import Foundation
+import UIKit
+
+@objc(TSCStormLanguageController)
+/// A controller that handles loading language files for Storm and provides methods to look up localisation strings.
+/// To understand how this controller works you must first understand the difference between a Storm Locale and a `Locale`.
+/// Storm Locales come in the format of "gbr_en" (Region, Language)
+/// `Locale` comes in the format of "en_GB" (Language, Region)
+/// `Locale` is able to ingest locales in the three letter format provided they are in the language_region format.
+/// This controller often re-organises the Storm file names to be in the language_region format before converting to `Locale`, once these are converted to `Locale` they can easily be compared with `Locale`s from the users device to find a match.
+public class StormLanguageController: NSObject {
+    
+    @objc(sharedController)
+    public static let shared = StormLanguageController()
+    
+    /// The dictionary of keys and values used for looking up language values for localisations.
+    var languageDictionary: [AnyHashable: Any]?
+    
+    /// The current language identifier
+    public var currentLanguage: String?
+    
+    /// The users language that they have forced as an overide. Usually different from the current device locale
+    public var overrideLanguage: TSCLanguage?
+    
+    /// The locales that the user prefers to view content in.
+    private var preferredLocales: [Locale]? {
+        
+        //Generate our preferred Locales based on the users preferences
+        var _preferredLocales = Locale.preferredLanguages.flatMap({ (languageString: String) -> Locale in
+            return Locale(identifier: languageString)
+        })
+        
+        //Add override locales if they exist
+        if let overrideObject = UserDefaults.standard.object(forKey: "TSCLanguageOverride") as? Data, let _overrideLanguage = NSKeyedUnarchiver.unarchiveObject(with: overrideObject) as? TSCLanguage, let localeIdentifier = _overrideLanguage.languageIdentifier {
+            _preferredLocales.insert(Locale(identifier: localeIdentifier), at: 0)
+            overrideLanguage = _overrideLanguage
+        }
+        
+        return _preferredLocales
+    }
+    
+    /// The locales that are available in the language packs
+    private var availableLocales: [LanguagePack]? {
```

Comment:
Would it not make more sense to name this `availableLanguagePacks` as order not to confuse it as an array of `Locale` objects?

Reply:
Good point I agree, I'll make the change now